src/CatalogService.Core/Entities/Category.cs

```diff
@@ -1,5 +1,101 @@
-namespace CatalogService.Core.Entities;
-
-public class Category {
-    
+using CatalogService.Core.Exceptions;
+using CatalogService.Core.Interfaces;
+using CatalogService.Core.Models;
+using FluentValidation;
+using OneOf;
+using OneOf.Types;
+
+namespace CatalogService.Core.Entities;
+
+public class Category {
+    private readonly IValidator<CategoryModel> _validator;
+    private readonly ICategoryRepository _repository;
+    private readonly CancellationToken _cancellationToken;
+
+    public Category(IValidator<CategoryModel> validator, ICategoryRepository repository, CancellationToken cancellationToken = default)
+    {
+        _validator = validator;
+        _repository = repository;
+        _cancellationToken = cancellationToken;
+    }
+
+    public async Task<OneOf<Success, CategoryFailureException>> Add(CategoryModel category)
+    {
+        var validate = _validator.Validate(category);
+
+        if (!validate.IsValid)
+            return new CategoryFailureException(validate.Errors);
+
+        try
+        {
+            await _repository.Add(category, _cancellationToken);
+        }
+        catch (Exception ex)
+        {
+            return new CategoryFailureException(ex.Message);
+        }
+
+        return new Success();
+    }
+
+    public async Task<OneOf<Success, NotFound, CategoryFailureException>> Update(CategoryModel category)
+    {
+        var validate = _validator.Validate(category);
+
+        if (!validate.IsValid)
+            return new CategoryFailureException(validate.Errors);
+
+        try
+        {
+            var response = await GetByName(category.Name);
+            if (response.IsT1)

```

Comment:
What will happen if `IsT1` will be something different than it previously returned from GetByName?

Reply:
In this case that's not possible, IsT1 means that the GetByName method returned None, which means that a record with that Name wasn't found on the DB, The only possibility for that method to return something other than the name specified would be it an SQL injection was possible, which is not the case as Dapper maps that threat when using parameters for queries.
It could be that the logic is confusing because IsT1 doesn't say much, I'll create a method for simplify that.